File: java/src/main/java/com/notifyservice/NotifyClient.java
package com.notifyservice;

import com.notifyservice.api.ApiClient;
import com.notifyservice.api.api.DiscordApi;
import com.notifyservice.api.api.InboxApi;
import com.notifyservice.api.api.NotificationsApi;
import com.notifyservice.api.api.PreferencesApi;
import com.notifyservice.api.api.SlackApi;
import com.notifyservice.api.api.SmsApi;
import com.notifyservice.api.api.TeamsApi;
import com.notifyservice.api.api.TemplatesApi;
import com.notifyservice.api.api.UsageApi;
import com.notifyservice.api.api.WebhooksApi;
import com.notifyservice.api.api.WorkflowsApi;
import java.net.http.HttpClient;

/**
 * Top-level NotifyService client: bearer auth, 429 retry, grouped tag resources.
 * Idempotency keys are passed via the generated send methods' Idempotency-Key
 * header parameter.
 */
public final class NotifyClient {

    private final ApiClient apiClient;

    public final NotificationsApi notifications;
    public final TemplatesApi templates;
    public final UsageApi usage;
    public final PreferencesApi preferences;
    public final WorkflowsApi workflows;
    public final WebhooksApi webhooks;
    public final InboxApi inbox;
    public final SmsApi sms;
    public final SlackApi slack;
    public final TeamsApi teams;
    public final DiscordApi discord;

    public NotifyClient(String baseUrl, String apiKey) {
        this(baseUrl, apiKey, 3);
    }

    public NotifyClient(String baseUrl, String apiKey, int maxRetries) {
        HttpClient retrying = new RetryingHttpClient(HttpClient.newHttpClient(), maxRetries);

        this.apiClient = new ApiClient();
        this.apiClient.setHttpClientBuilder(new SingletonHttpClientBuilder(retrying));
        this.apiClient.updateBaseUri(stripTrailingSlash(baseUrl));
        this.apiClient.setRequestInterceptor(builder -> builder.header("Authorization", "Bearer " + apiKey));

        this.notifications = new NotificationsApi(apiClient);
        this.templates = new TemplatesApi(apiClient);
        this.usage = new UsageApi(apiClient);
        this.preferences = new PreferencesApi(apiClient);
        this.workflows = new WorkflowsApi(apiClient);
        this.webhooks = new WebhooksApi(apiClient);
        this.inbox = new InboxApi(apiClient);
        this.sms = new SmsApi(apiClient);
        this.slack = new SlackApi(apiClient);
        this.teams = new TeamsApi(apiClient);
        this.discord = new DiscordApi(apiClient);
    }

    private static String stripTrailingSlash(String url) {
        return url.endsWith("/") ? url.substring(0, url.length() - 1) : url;
    }
}

File: php/src/NotifyClient.php
<?php

declare(strict_types=1);

namespace NotifyService;

use GuzzleHttp\Client as GuzzleClient;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Middleware;
use NotifyService\Api\Api\DiscordApi;
use NotifyService\Api\Api\InboxApi;
use NotifyService\Api\Api\NotificationsApi;
use NotifyService\Api\Api\PreferencesApi;
use NotifyService\Api\Api\SlackApi;
use NotifyService\Api\Api\SMSApi;
use NotifyService\Api\Api\TeamsApi;
use NotifyService\Api\Api\TemplatesApi;
use NotifyService\Api\Api\UsageApi;
use NotifyService\Api\Api\WebhooksApi;
use NotifyService\Api\Api\WorkflowsApi;
use NotifyService\Api\Configuration;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\ResponseInterface;

final class NotifyClient
{
    public readonly NotificationsApi $notifications;
    public readonly TemplatesApi $templates;
    public readonly UsageApi $usage;
    public readonly PreferencesApi $preferences;
    public readonly WorkflowsApi $workflows;
    public readonly WebhooksApi $webhooks;
    public readonly InboxApi $inbox;
    public readonly SMSApi $sms;
    public readonly SlackApi $slack;
    public readonly TeamsApi $teams;
    public readonly DiscordApi $discord;

    public function __construct(string $baseUrl, string $apiKey, int $maxRetries = 3)
    {
        $config = Configuration::getDefaultConfiguration()
            ->setHost(rtrim($baseUrl, '/'))
            ->setAccessToken($apiKey);

        $stack = HandlerStack::create();
        $stack->push(self::retryMiddleware($maxRetries));
        $http = new GuzzleClient(['handler' => $stack]);

        $this->notifications = new NotificationsApi($http, $config);
        $this->templates = new TemplatesApi($http, $config);
        $this->usage = new UsageApi($http, $config);
        $this->preferences = new PreferencesApi($http, $config);
        $this->workflows = new WorkflowsApi($http, $config);
        $this->webhooks = new WebhooksApi($http, $config);
        $this->inbox = new InboxApi($http, $config);
        $this->sms = new SMSApi($http, $config);
        $this->slack = new SlackApi($http, $config);
        $this->teams = new TeamsApi($http, $config);
        $this->discord = new DiscordApi($http, $config);
    }

    public static function retryMiddleware(int $maxRetries): callable
    {
        return Middleware::retry(
            static function (int $retries, RequestInterface $request, ?ResponseInterface $response = null) use ($maxRetries): bool {
                if ($retries >= $maxRetries || $response === null) {
                    return false;
                }
                $code = $response->getStatusCode();
                return $code === 429 || $code >= 500;
            },
            static function (int $retries, ?ResponseInterface $response = null): int {
                if ($response !== null && $response->hasHeader('Retry-After')) {
                    return max(0, (int) $response->getHeaderLine('Retry-After')) * 1000;
                }
                return (int) (500 * (2 ** $retries));
            }
        );
    }
}
